guides: protocol summary

pinheadmz opened the pull request:
Adds a cheatsheet for the Handshake protocol as a guide.

Covers basic data about the blockchain, monetary supply, covenant rules, auction rules, airdrops, claims, new script OP codes and more!

An excellent starting to point to anyone familiar with the Bitcoin protocol, that has five minutes to learn everything they can about Handshake.

tynes commented on src/guides/protocol.md:
```diff
@@ -0,0 +1,230 @@
+[4000000 Weight Units](https://github.com/handshake-org/hsd/blob/b32c27bbb05e69b86c0e80aea3ecc394c5589186/lib/protocol/consensus.js#L176-L177)
+
+- Hash algorithm:
+[BLAKE2b256 (sha3) keyed with KMAC256](https://github.com/handshake-org/hsd/blob/b32c27bbb05e69b86c0e80aea3ecc394c5589186/lib/mining/mine.js#L28-L29)
```
jj has been calling this `cBlake`

pinheadmz replied:
👍

tynes commented on src/guides/protocol.md:
```diff
@@ -0,0 +1,230 @@
+A witness program can encode a public key or a script hash, but those keys and
+scripts will appear in transaction _inputs_.
+
+Handshake also introduces a new field in transaction outputs called a `covenant`.
```
Maybe useful to copy this from the white paper:

``` c
struct output_s {
  int64_t value;
  uint8_t script[];
  struct covenant_s {
    uint8_t action;
    uint8_t arguments[][];
  } covenant;
} output;
```

pinheadmz replied:
The javascript version of that is linked in the next line

tynes replied:
The covenant type is obscured behind the `covenant.write` method

pinheadmz replied:
True, but a few lines down I link to the `write` method in covenant.js. What I want to show here is more relevant to transaction serialization, then under covenants section more details are revealed.

tynes replied:
Ok I'm not super opinionated on this but I do think that seeing the struct is interesting

tynes commented on src/guides/protocol.md:
```diff
@@ -0,0 +1,230 @@
+
+Un-reserved names are not all available immediately when mainnet launches. They
+become available over the first year, based on a simple algorithm that
+[compares the name's hash against a 52-week modulus.](https://github.com/handshake-org/hsd/blob/b32c27bbb05e69b86c0e80aea3ecc394c5589186/lib/covenants/rules.js#L274-L295)
```
Maybe mention that you can use the RPC `getnameinfo` to determine when the name will be available

pinheadmz replied:
Sure, wouldn't hurt. I just wanted to focus more abstractly on the protocol in this doc rather than usage, but I'll add this one.

tynes commented on src/guides/protocol.md:
```diff
@@ -0,0 +1,230 @@
+no signature is required to generate their airdrop transactions (of course a signature
+is required to SPEND from the airdrop, like any other address). Because these transactions
+can be generated and submitted by any node at any time, their
+[fees are hard-coded.](https://github.com/handshake-org/hsd/blob/b32c27bbb05e69b86c0e80aea3ecc394c5589186/lib/primitives/airdropproof.js#L18-L19)
```
fees are hard-coded and enforced by consensus

pinheadmz replied:
👍

tynes commented on src/guides/protocol.md:
```diff
@@ -0,0 +1,230 @@
+| < 720                            | ~5 days      | `BIDDING`         |
+| < 1,440                          | ~10 days     | `REVEAL`          |
+| > 0                              |              | `CLOSED`          |
+
```
Maybe mention that this is for mainnet and that the values for each of the networks can be found at `lib/protocol/networks.js`

pinheadmz replied:
This is mentioned at the very top of the article!

tynes replied:
`all values below describe mainnet parameters.`

Gotcha 👍

tynes commented on src/guides/protocol.md:
```diff
@@ -0,0 +1,230 @@
+after the reveal period ends.
+
+Names must be renewed within
+[two years](https://github.com/handshake-org/hsd/blob/b32c27bbb05e69b86c0e80aea3ecc394c5589186/lib/protocol/networks.js#L250)
```
Good find, I had thought it was 1 year.  The max number of names that could be theoretically supported by the blockchain is if every block is full of only renewal transactions

pinheadmz replied:
👍

pinheadmz commented:
@tynes addressed your comments and force-pushed without all the build crap, a little easier to read now :-)

tynes commented on src/guides/protocol.md:
```diff
@@ -0,0 +1,233 @@
+    - [Airdrops](https://github.com/handshake-org/hsd/blob/b32c27bbb05e69b86c0e80aea3ecc394c5589186/lib/mining/template.js#L310-L324)
+    and [Claims](https://github.com/handshake-org/hsd/blob/b32c27bbb05e69b86c0e80aea3ecc394c5589186/lib/mining/template.js#L282-L308)
+    are included in the coinbase transaction. Proofs for both of these types are
+    [verified during block contextual checks, in `tx.js`.](https://github.com/handshake-org/hsd/blob/b32c27bbb05e69b86c0e80aea3ecc394c5589186/lib/primitives/tx.js#L418-L472)
```
Nit: `lib/primitives/tx.js` over `tx.js`

pinheadmz replied:
👍

tynes commented on src/guides/protocol.md:
```diff
@@ -0,0 +1,233 @@
+
+- Base unit: [dollarydoo](https://github.com/handshake-org/hsd/blob/b32c27bbb05e69b86c0e80aea3ecc394c5589186/lib/pkg.js#L52)
+
+- 1 Coin = [1,000,000 dollarydoos](https://github.com/handshake-org/hsd/blob/b32c27bbb05e69b86c0e80aea3ecc394c5589186/lib/protocol/consensus.js#L30)
```
1 Coin vs 1 HNS ?

pinheadmz replied:
👍

pinheadmz commented:
Updated again with hash OP codes and SIGHASH_SINGLE.

I was thinking, this file could be added easily to `hsd/docs` as well (or instead of a guide) ?

tynes commented:
I think another cool thing to add would be the hardening soft fork to make it so that airdrops with weak keys can no longer be claimed

pinheadmz commented:
Thanks @tynes I added a blob about the soft fork: https://github.com/handshake-org/handshake-org.github.io/blob/bdb9647f3df8788d15cdc6aa8f0b0565e35e21ac/src/guides/protocol.md#rsa-1024-soft-fork

Squashed

tynes commented:
This looks good to me

kilpatty commented:
This is awesome, looks great!

tynes commented on src/guides/protocol.md:
```diff
@@ -0,0 +1,271 @@
+[4000000 Weight Units](https://github.com/handshake-org/hsd/blob/b32c27bbb05e69b86c0e80aea3ecc394c5589186/lib/protocol/consensus.js#L176-L177)
+
+- Hash algorithm: "cBLAKE" --
+[BLAKE2b256 (sha3) keyed with KMAC256](https://github.com/handshake-org/hsd/blob/b32c27bbb05e69b86c0e80aea3ecc394c5589186/lib/mining/mine.js#L28-L29)
```
BLAKE2b256 is not sha3, the sha3 part comes from KMAC256

pinheadmz replied:
I see, thanks - so maybe better phrasing is `BLAKE2b256 keyed with KMAC256 (sha3)` ?

tynes replied:
That sounds good

tynes commented on src/guides/protocol.md:
```diff
@@ -0,0 +1,271 @@
+- Hash algorithm: "cBLAKE" --
+[BLAKE2b256 (sha3) keyed with KMAC256](https://github.com/handshake-org/hsd/blob/b32c27bbb05e69b86c0e80aea3ecc394c5589186/lib/mining/mine.js#L28-L29)
+
+- Segregated Witness: Handshake is 100% SegWit. No soft fork, no technical debt.
```
I'd remove the `No soft fork, no technical debt.` as that sounds antagonistic

tynes commented on src/guides/protocol.md:
```diff
@@ -0,0 +1,271 @@
+
+- Coinbase transactions:
+
+    - Because Handshake is 100% SegWit, there are no `ScriptSig`'s, just witness data.
```
No need for `'` in `ScriptSig`'s

pinheadmz commented:
@tynes thanks, last 3 nits addressed and squashed

pinheadmz commented on src/guides/protocol.md:
```diff
@@ -0,0 +1,271 @@
+- Block interval time:
+[10 minutes](https://github.com/handshake-org/hsd/blob/b32c27bbb05e69b86c0e80aea3ecc394c5589186/lib/protocol/networks.js#L166)
+
+- PoW Difficulty adjustment interval: [144 Blocks (one day)](https://github.com/handshake-org/hsd/blob/b32c27bbb05e69b86c0e80aea3ecc394c5589186/lib/protocol/networks.js#L158-L159)
```
PoW actually retargets every single block, but it's BASED on the last 144 blocks as a target for one day of time.
